### ISSUE TITLE
Morphing between two instances of a component leaks proxy nodes

### ISSUE THREAD

**Issue by bcomnes** [choojs/nanocomponent#29]
I put together an example of nanocomponent leaking proxy elements into the real DOM when morphing between two instances of the same component:

https://gist.github.com/bcomnes/fa4fa3a83807b4189356699c17207a0b

Two issues at play here:

- You need to set the caller of `on-load`, an [undocumented argument](https://github.com/shama/on-load/blob/master/index.js#L69) that is critical to firing lifecycle methods correctly. (see https://github.com/hypermodules/cache-component/blob/master/index.js#L40 for an example)
- The real DOM reference is not properly set on the first morph of the second instance.  Subsequent visits will present a failing isSameNode test that results in leaked proxy nodes into the real dom. A possible fix is to set internal instance ID's but I haven't had a chance to bench how that affects performance: https://github.com/hypermodules/cache-component/blob/master/index.js#L49

Want me to PR these fixes to nanocomponent?

**Comment by yoshuawuyts** [comment]
> You need to set the caller of on-load, an undocumented argument that is critical to firing lifecycle methods correctly. (see https://github.com/hypermodules/cache-component/blob/master/index.js#L40 for an example)

Could you perhaps explain why this is critical? I believe you that it is, but I'm not sure I follow haha

---

> The real DOM reference is not properly set on the first morph of the second instance. Subsequent visits will present a failing isSameNode test that results in leaked proxy nodes into the real dom. A possible fix is to set internal instance ID's but I haven't had a chance to bench how that affects performance: https://github.com/hypermodules/cache-component/blob/master/index.js#L49

Oh yeah, you're right - so what I suggest we do here is set an `id` property if there isn't one already. and then do the same check you proposed. That should take care of it all (:

**Comment by yoshuawuyts** [comment]
From IRC:
```txt
01:49 <louisc> acorn throws the same error
01:54 <yoshuawuyts> bret: lol, you've found a super interesting edge case; I'm not sure how to fix it haha
01:55 <yoshuawuyts> bret: like: your solution accounts for morphing component -> component and making sure it comes through
01:56 <yoshuawuyts> bret: but what if the reference is lost in a regular morph
01:57 <yoshuawuyts> bret: oh no nevermind, the element onload id will still be transferred and it'll work lol
01:58 <bret> I don't even fully understand what's going on tbh. I think have a rough idea though
02:01 <yoshuawuyts> bret: no cool, you're totally right on the second instance
02:01 <yoshuawuyts> bret: I still don't completely follow on the first thing tho, what's the "caller" argument for?
02:02 <yoshuawuyts> ohhh right
02:02 <yoshuawuyts> now I see lol
02:02 <yoshuawuyts> bret: ok ok, replied to your thing - you're right in both cases. PRs super welcome!
02:03 <yoshuawuyts> bret: think setting an "id" by default is the right way to go, that way we don't have to do stuff like ccid and can also benefit from https://github.com/yoshuawuyts/nanomorph/pull/61
02:03 <bret> Good idea
02:03 <yoshuawuyts> bret: thanks heaps for tracking this down! :D :D
02:04 <bret> Does that not polite the global namespace though?
02:04 <bret> Pollute
02:04 <louisc> yoshuawuyts: is there some kind of node debug thing where I can watch bankai step through the code to see at what particular point it chokes?
02:04 <bret> Ids get hoisted to global cars I think
02:04 <bret> Vars
02:05 <yoshuawuyts> bret: it does, but that's cool I think - have never heard of perf issues regarding Ids
02:05 <bret> K
02:07 <yoshuawuyts> bret: using a similar scheme as onload has should work fine I think
```

**Comment by bcomnes** [comment]
Closed by 6.0.0
